src/jit.c

```diff
@@ -1278,92 +1277,84 @@ static void prepare_translate(struct jit_state *state)
 }
 
 
-static int count = 0;
-static int reg_table[32];
-static int vm_reg[3] = {0}; /* enum x64_reg/a64_reg */
+static int vm_reg[3]; /* enum x64_reg/a64_reg */
 
 static void reset_reg()
 {
-    count = 0;
-    for (int i = 0; i < N_RV_REGS; i++) {
-        reg_table[i] = -1;
-    }
     for (int i = 0; i < n_host_regs; i++) {
-        register_map[i].dirty = 0;
-    }
-}
-
-/* Find the vm register index that mapped to the given host register. */
-static inline int find_rv_reg(int reg_idx)
-{
-    for (int i = 0; i < N_RV_REGS; i++) {
-        if (reg_table[i] == reg_idx)
-            return i;
+        register_map[i].vm_reg_idx = -1;
+        register_map[i].dirty = false;
+        register_map[i].alive = false;
     }
-    return -1;
 }
 
 /* Save host register if it is dirty. */
-static inline void save_reg(struct jit_state *state, int reg_idx)
+static inline void save_reg(struct jit_state *state, int idx)
 {
-    for (int i = 0; i < n_host_regs; i++) {
-        /* ignore nonvolatile and parameter registers */
-        if (register_map[i].reg_idx != reg_idx)
-            continue;
-        if (!register_map[i].dirty)
-            continue;
+    assert(idx > -1 && idx < n_host_regs);
 
-        assert(find_rv_reg(reg_idx) != -1);
-        emit_store(state, S32, reg_idx, parameter_reg[0],
-                   offsetof(riscv_t, X) + 4 * find_rv_reg(reg_idx));
-        register_map[i].dirty = 0;
+    if (!register_map[idx].dirty)
         return;
-    }
+
+    emit_store(state, S32, register_map[idx].reg_idx, parameter_reg[0],
+               offsetof(riscv_t, X) + 4 * register_map[idx].vm_reg_idx);
+    register_map[idx].dirty = 0;
 }
 
 static void store_back(struct jit_state *state)
 {
-    for (int i = 0; i < N_RV_REGS; i++) {
-        if (reg_table[i] == -1)
+    for (int i = 0; i < n_host_regs; i++) {
+        if (register_map[i].vm_reg_idx == -1)
             continue;
-        save_reg(state, reg_table[i]);
+        save_reg(state, i);
     }
 }
 
 /* Unmap the vm register to the host register. */
-static inline void unmap_vm_reg(int reg_idx)
+static inline void unmap_vm_reg(int idx)
 {
-    for (uint32_t i = 0; i < N_RV_REGS; i++) {
-        if (reg_table[i] != reg_idx)
-            continue;
-        reg_table[i] = -1;
-        return;
-    }
+    /* check dirty before unmap */
+    assert(idx > -1 && idx < n_host_regs);
+    register_map[idx].vm_reg_idx = -1;
 }
 
-static inline void set_reg_table(int vm_reg_idx, int reg_idx)
+#include "reg-alloc.c"
```

Comment:
Since the file `reg-alloc.c` is used only in the tier 1 JIT compiler and is relatively shorter, it would make sense to move the implementation directly into file `jit.c`.